### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by elenapetra**
endpoints:
  GET - https://zq3zdypgs7.execute-api.eu-west-1.amazonaws.com/dev/products
  GET - https://zq3zdypgs7.execute-api.eu-west-1.amazonaws.com/dev/products/{productId}
  POST - https://zq3zdypgs7.execute-api.eu-west-1.amazonaws.com/dev/products

There were no changes made on FE for the current task.

### CODE REVIEW

File: product-service/serverless.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    events:
+      - http:
+          path: products
+          method: get
+          cors: true
+    environment:
+      PRODUCTS_TABLE: "products-db"
+      STOCKS_TABLE: "stocks-db"
+
+  getProductsById:
```
Comment (dav-dio) [inline product-service/serverless.yml]: This should be getProductById since you're querying a specific product

File: product-service/handler.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        (item) => item.product_id === product.id
+      );
+      return {
+        ...product,
+        count: stockInfo ? stockInfo.count : 0,
+      };
+    });
+    return {
+      statusCode: 200,
+      headers: {
```
Comment (dav-dio) [inline product-service/handler.js]: Doesn't the cors setting in serverless.yml for this endpoint take care of this? Or how come you need these headers in this response but not the others?

File: product-service/handler.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      dynamodb.scan(stocksParams).promise(),
+    ]);
+
+    const productsWithStocks = productsData.Items.map((product) => {
+      const stockInfo = stocksData.Items.find(
+        (item) => item.product_id === product.id
+      );
+      return {
+        ...product,
+        count: stockInfo ? stockInfo.count : 0,
```
Comment (dav-dio) [inline product-service/handler.js]: You're assuming count will exist on stockInfo if stockInfo exists.
You could do something safer like
```suggestion
        count: stockInfo?.count ?? 0,
```

Then again, if the intent is for this code to throw if for some reason you get stockInfo back but it doesn't have a count property (such as if the data schema has changed), this is fine

File: product-service/src/utils/generateUniqueId.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const crypto = require("crypto");
+
+function generateUniqueId() {
```
Comment (dav-dio) [inline product-service/src/utils/generateUniqueId.js:3]: You could have just used the uuid Node library, but this is fine I guess
https://www.npmjs.com/package/uuid?activeTab=readme

File: product-service/handler.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      },
+    };
+
+    await Promise.all([
+      dynamodb.put(productParams).promise(),
+      dynamodb.put(stockParams).promise(),
+    ]);
+
+    const responseData = {
+      data: {
```
Comment (dav-dio) [inline product-service/handler.js:119]:
```suggestion
      data: {
      ...productParams.Item, count: data.count
      }
```

File: product-service/serverless.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  iam:
+    role:
+      statements:
+        - Effect: Allow
+          Action:
+            - dynamodb:Scan
+            - dynamodb:GetItem
+            - dynamodb:PutItem
+          Resource:
+            - "arn:aws:dynamodb:eu-west-1:533267237683:table/products-db"
```
Comment (dav-dio) [inline product-service/serverless.yml:18]: You could use global environment variables for the table names (region, account etc.) that you could then reference here as well as in your individual endpoints.
Cf. https://www.serverless.com/framework/docs/providers/aws/guide/variables#reference-properties-in-serverlessyml

**Review by dav-dio — APPROVED** [review]

File: product-service/serverless.yml (hunk 1 above)
Comment (elenapetra) [reply product-service/serverless.yml]: You're right, I've changed it to getProductById.

File: product-service/src/utils/generateUniqueId.js (hunk 4 above)
Comment (elenapetra) [reply product-service/src/utils/generateUniqueId.js:3]: I've tried to use uuid module but I was getting the error: "Cannot find module 'uuid' " so I replaced it with crypto.
Later I found out that I can add it as a layer to the lambda function.

File: product-service/handler.js (hunk 5 above)
Comment (elenapetra) [reply product-service/handler.js:119]: I've applied the suggested changes. Thanks.

File: product-service/handler.js (hunk 2 above)
Comment (elenapetra) [reply product-service/handler.js]: You are right, I don't need it. I was getting a CORS error in browser's console, but now it's gone.

File: product-service/handler.js (hunk 3 above)
Comment (elenapetra) [reply product-service/handler.js]: I will go with the option you've suggested, as it's safer.

File: product-service/serverless.yml (hunk 6 above)
Comment (elenapetra) [reply product-service/serverless.yml:18]: I've defined environment variables globally.

**Review by dav-dio — APPROVED** [review]

**Pull request merged**